### PR TITLE
feat!: `no-unused-vars` default caughtErrors to 'all'

### LINKED ISSUE #17974: Rule Change: Change default for no-unused-var's caughtErrors option to "all"
### What rule do you want to change?

no-unused-vars

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
try {
  // ...
} catch (error) {
  // ...
}
```

### What does the rule currently do for this code?

[`no-unused-var`'s `caughtErrors` option](https://eslint.org/docs/latest/rules/no-unused-vars#caughterrors) defaults to `"none"`. It won't report on an unused caught error by default.

### What will the rule do after it's changed?

With `caughtErrors` defaulting to `"all"`, it would report by default.

Note that this issue just suggests changing the default value for the option. It does not suggest changing any rule logic around naming ignore patterns or removing the option.

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

Optional catch bindings were added to ECMAScript 2019 in https://tc39.es/proposal-optional-catch-binding. Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch#browser_compatibility they're supported in Node.js >= 10.0.0 and evergreen browsers starting in 2020 (Edge) or <=2018 (others). Given that ESLint's minimum Node version is 12, I think compatibility with older runtimes is no longer a blocker for this issue?

This would be a breaking change. Pity I only thought to request it _after_ the v9 alphas were set in motion 😄. Noticed in https://github.com/typescript-eslint/typescript-eslint/issues/8224.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Changes the default value for the `no-unused-vars` rule's `caughtErrors` option from `"none"` to `"all"`.

Fixes #17974.

### CODE REVIEW

**Comment by Tanujkanti4441** [comment]
Could you change the docs too!

**Comment by JoshuaKGoldberg** [comment]
D'oh, right, thanks.

**Comment by snitin315** [comment]
We need to mention this in the migration guide, where we have mentioned other breaking changes as well.

File: docs/src/rules/no-unused-vars.md [hunk 1]
```diff
@@ -283,20 +283,24 @@ The `caughtErrors` option is used for `catch` block arguments validation.
 
 It has two settings:
 
-* `none` - do not check error objects. This is the default setting.
-* `all` - all named arguments must be used.
+* `all` - all named arguments must be used. This is the default setting.
+* `none` - do not check error objects.
 
-#### caughtErrors: none
+:::
```
Comment (mdjermanovic) [inline docs/src/rules/no-unused-vars.md:289]: I think this `:::`  should be moved to the end of the `#### caughtErrors: none` section where it's missing.

File: docs/src/use/migrate-to-9.0.0.md [hunk 2]
```diff
@@ -362,6 +363,33 @@ if (test) {
...
+
+```js
+/*eslint no-unused-vars: "error"*/
+try {}
+catch (error) {
+    // 'error' is defined but never used
+}
+```
+
+**To address:** If you want to allow unused caught errors, such as when writing code that will be directly run in a environment that does not support ES2019 optional catch bindings, set the `caughtErrors` option to `"none"`.
```
Comment (mdjermanovic) [inline docs/src/use/migrate-to-9.0.0.md:380]:
```suggestion
**To address:** If you want to allow unused caught errors, such as when writing code that will be directly run in an environment that does not support ES2019 optional catch bindings, set the `caughtErrors` option to `"none"`.
```

File: docs/src/use/migrate-to-9.0.0.md [hunk 3]
```diff
@@ -33,6 +33,7 @@ The lists below are ordered roughly by the number of users each change is expect
 * [`no-restricted-imports` now accepts multiple config entries with the same `name`](#no-restricted-imports)
 * [`"eslint:recommended"` and `"eslint:all"` strings no longer accepted in flat config](#string-config)
 * [`no-inner-declarations` has a new default behavior with a new option](#no-inner-declarations)
+* [`no-unused-vars` now defaults `caughtErrors` to `"all"`](#no-unused-vars)
```
Comment (bmish) [inline docs/src/use/migrate-to-9.0.0.md:36]: As an aside, it would be nice in a separate PR to group all default rule option changes together in this list or a sub-list.

File: docs/src/use/migrate-to-9.0.0.md [hunk 4]
```diff
@@ -362,6 +363,33 @@ if (test) {
 
 **Related issue(s):** [#15576](https://github.com/eslint/eslint/issues/15576)
 
+## <a name="no-unused-vars"></a> `no-unused-vars` now defaults `caughtErrors` to `"all"`
+
+ESLint v9.0.0 changes the default value for the `no-unused-vars` rule's `caughtErrors` option.
+Previously it defaulted to `"none"` to never check whether caught errors are used.
```
Comment (amareshsm) [inline docs/src/use/migrate-to-9.0.0.md:369]:
```suggestion
Previously it defaulted to `"none"` to never check whether caught errors were used.
```

**Review by nzakas — APPROVED** [review]
LGTM pending @amareshsm's grammar update.

**Review by amareshsm — APPROVED** [review]
LGTM.
